src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Allows limited reuse of heap buffers to improve memory pressure. This cache does not ensure
+    /// that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
+    {
+        private readonly ulong _minSize;
+        private readonly ulong _maxSize;
+
+        // Want to be able to use in low level libraries so we'll keep an array
+        private SafeHeapHandle[] _buffers;
+        private uint _bufferCount;
+        private readonly object _bufferLock = new object();
+
+        /// <param name="minSize">Smallest buffer size to allocate in bytes.</param>
+        /// <param name="maxSize">The largest buffer size to cache in bytes.</param>
+        /// <param name="maxBuffers">The maximum number of buffers to cache.</param>
+        public SafeHeapHandleCache(ulong minSize = 64, ulong maxSize = 1024 * 4, uint maxBuffers = 0)
+        {
+            _minSize = minSize;
+            _maxSize = maxSize;
+            maxBuffers = maxBuffers > 0 ? maxBuffers : (uint)Environment.ProcessorCount * 4;
+            _buffers = new SafeHeapHandle[maxBuffers];
+        }
+
+        /// <summary>
+        /// Get a HeapHandle
+        /// </summary>
+        public SafeHeapHandle Acquire(ulong minSize = 0)
+        {
+            if (minSize < _minSize) minSize = _minSize;
+
+            SafeHeapHandle buffer = null;
+            lock (_bufferLock)
+            {
+                if (_bufferCount > 0) buffer = _buffers[--_bufferCount];
+            }
+
+            if (buffer == null)
+                buffer = new SafeHeapHandle(minSize);
+            else if (buffer.ByteLength < minSize)
+                buffer.Resize(minSize);
```

Comment:
I didn't because it frequently expands in place and I expect that pressure on size is something we want to grow with. Can always revisit.

I've talked with both @sokket and @KrzysztofCwalina. @sokket and I had a long talk about it and are trying to track some sort of convergence if possible.